pyo3-ffi/src/pystate.rs

```diff
@@ -73,9 +73,55 @@ pub enum PyGILState_STATE {
     PyGILState_UNLOCKED,
 }
 
+struct HangThread;
+
+impl Drop for HangThread {
+    fn drop(&mut self) {
+        loop {
+            #[cfg(target_family = "unix")]
+            unsafe {
+                libc::pause();
+            }
+            #[cfg(not(target_family = "unix"))]
+            std::thread::sleep(std::time::Duration::from_secs(9_999_999));
+        }
+    }
+}
+
+// The PyGILState_Ensure function will call pthread_exit during interpreter shutdown,
+// which causes undefined behavior. Redirect to the "safe" version that hangs instead,
+// as Python 3.14 does.
+//
+// See https://github.com/rust-lang/rust/issues/135929
+
+// C-unwind only supported (and necessary) since 1.71. Python 3.14+ does not do
+// pthread_exit from PyGILState_Ensure (https://github.com/python/cpython/issues/87135).
+mod raw {
+    #[cfg(all(not(Py_3_14), rustc_has_extern_c_unwind))]
+    extern "C-unwind" {
+        #[cfg_attr(PyPy, link_name = "PyPyGILState_Ensure")]
+        pub fn PyGILState_Ensure() -> super::PyGILState_STATE;
+    }
+
+    #[cfg(not(all(not(Py_3_14), rustc_has_extern_c_unwind)))]
+    extern "C" {
+        #[cfg_attr(PyPy, link_name = "PyPyGILState_Ensure")]
+        pub fn PyGILState_Ensure() -> super::PyGILState_STATE;
+    }
+}
+
+#[cfg(not(Py_3_14))]
+pub unsafe extern "C" fn PyGILState_Ensure() -> PyGILState_STATE {
+    let guard = HangThread;
+    let ret: PyGILState_STATE = raw::PyGILState_Ensure();
+    std::mem::forget(guard);
+    ret
+}
```

Comment:
Disclaimer: I'm really no expert here.

According to my understanding of the `C-unwind` this is still UB. Specifically [here](https://rust-lang.github.io/rfcs/2945-c-unwind-abi.html#forced-unwinding) is mentioned that forced unwinding (like `pthread_exit`) can not rely on `Drop` and thus crossing such framed is considered UB.

Given that this was UB before as well, but this one works, maybe that's better 🤷 (We really need a Python API to do this properly)

Reply:
So bjorn claimed this way is unspecified behavior but should work. I think it's better then the C++ version.

Reply:
https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/catching.20pthread_exit.20to.20hang.20from.20Rust

Reply:
Thanks, I think that makes sense to me. I guess we aren't guaranteed that `drop` will run, but in case it is not run, we are in the same situation as currently (presumably a crash). So given this (currently) works in practice it seems like a net win to me.

> I think it's better then the C++ version.

Definitely agree here, I was not really a fan of adding a C++ toolchain dependency 👍

Reply:
added a comment